### ISSUE TITLE
YoloV5 with OpenVino

### ISSUE THREAD

**Issue by MasIgor** [ultralytics/yolov5#6905]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello, sorry if this is not the correct place to ask this question, but I am not able to find a solution anywhere.

I am trying to run a yoloV5 Model with Openvino. To do so, i converted the model with the export script to openvino format.

the output of the openvino file is 1x25200x13

I start the Inference with the code below, and get my output variable with the about 25000 rows with 13 values each. Its my understanding that the first 5 values indicate the box size and location and the confidence that there is anything in this box, and then the last 8 are my classes. 

So far so good. But now that I have these 25.200 rows, how do I get my actual prediction? 

OpenVino guys told me that this issue is not related to their system as its a direct result of a conversion from the yolo model.

could you please help me out? I am sharing the code, and a zip with my yolo model, my onnx model, my openvino model and 2 sample images. 

Thank you!

My files: [www.igormasin.it/fileuploads/yolo.zip](url)

```
 ie = IECore()
        devices = ie.available_devices
        for device in devices:
            device_name = ie.get_metric(device_name=device, metric_name="FULL_DEVICE_NAME")
            print(f"{device}: {device_name}")
        classification_model_xml = "bests.xml"
        net = ie.read_network(model=classification_model_xml)
        exec_net = ie.load_network(network=net, device_name="CPU")
        input_layer = next(iter(net.input_info))
        output_layer = net.outputs['output']
        image_filename = "test.jpg"
        image = cv2.imread(image_filename)
        print(f"input layout: {net.input_info[input_layer].layout}")
        print(f"input precision: {net.input_info[input_layer].precision}")
        print(f"input shape: {net.input_info[input_layer].tensor_desc.dims}")
        print(f"output layout: {output_layer.layout}")
        print(f"output precision: {output_layer.precision}")
        print(f"output shape: {output_layer.shape}")
        N, C, H, W = net.input_info[input_layer].tensor_desc.dims
        print(N, C, H, W)
        image.shape
        input_data = np.expand_dims(np.transpose(image, (2, 0, 1)), 0).astype(np.float32)
        input_data.shape
        result = exec_net.infer({input_layer: input_data})
        output = result['output']
```

**Comment by glenn-jocher** [comment]
@tanzerlana you don't have to worry about any of those details, just follow the Usage examples displayed after OpenVINO export:
<img width="1504" alt="Screenshot 2022-03-08 at 22 22 51" src="https://user-images.githubusercontent.com/26833433/157327360-6a81493d-6a17-4369-83ea-f518390a7aa4.png">

**Comment by anujdutt9** [comment]
Hi @tanzerlana If I am understanding it correctly, you have the OpenVINO model for YOLOv5. Now you are trying to run inference using the OpenVINO API's on a CPU/GPU/MYRIAD device.

If this is correct, then your code looks fine. What you are missing is **Non-Maximum Suppression** code based on some 
 confidence threshold value. Since, your model predicts 25200 values, you'll need to suppress the outputs based on some confidence threshold that you can set, and all predictions below that threshold will be removed/suppressed.

Then you'll get the final outputs that you can plot on your image to visualize the bounding boxes.

Checkout the code here for reference: https://github.com/violet17/yolov5_demo

Hope this helps.

**Comment by MasIgor** [comment]
Hello @anujdutt9 , thank you very much! yes that is exactly what I am trying to do, since I do not have a GPU and the CPU is too slow I would like to use the internal GPU of my i7 to speed things up.

I have tried to run the demo (the one for version 2021.3) and get this result:

```
python.exe C:/Qt/workspace/KI_Time_Test/openvinoscr.py -i test.jpg -m bests.xml
[ INFO ] Creating Inference Engine...
[ INFO ] Loading network:
	bests.xml
[ INFO ] Preparing inputs
[ INFO ] Loading model to the plugin
[ INFO ] Starting inference...
To close the application, press 'CTRL+C' here or switch to the output window and press ESC key
To switch between sync/async modes, press TAB key in the output window
[ INFO ] Layer 345 parameters: 
[ INFO ]          classes : 8
[ INFO ]          num     : 3
[ INFO ]          coords  : 4
[ INFO ]          anchors : [10.0, 13.0, 16.0, 30.0, 33.0, 23.0, 30.0, 61.0, 62.0, 45.0, 59.0, 119.0, 116.0, 90.0, 156.0, 198.0, 373.0, 326.0]
Traceback (most recent call last):
  File "C:/Qt/workspace/KI_Time_Test/openvinoscr.py", line 412, in <module>
    sys.exit(main() or 0)
  File "C:/Qt/workspace/KI_Time_Test/openvinoscr.py", line 332, in main
    args.prob_threshold)
  File "C:/Qt/workspace/KI_Time_Test/openvinoscr.py", line 162, in parse_yolo_region
    out_blob_n, out_blob_c, out_blob_h, out_blob_w = blob.shape
ValueError: too many values to unpack (expected 4)

Process finished with exit code 1

```

Just so I understand it correctly: (back to my code posted above) the fifth value in the result array is the confidence that there is something in there.. correct? so the non-maximum-suppression would remove all values that are below a certain threshold... I have run a maxarg on vertical axis and the highest value in there is about 0.3, and the single classes have pretty low values too in that particular row. Does it mean that it didnt find anything? with the yolo model it finds everything with very high confidence...

Also, could you please explain what this letterboxing is? I have read it multiple times, but couldnt figure out what it does. 

Thank you cor your help! very appreciated!

**Comment by MasIgor** [comment]
@anujdutt9 I found the problem... Exporting it directly as openvino does not work. If I export it to onnx, and then specify the nodes like shown in the sample you linked then it works. thank you very much! the problem I am facing now is that in both cases, (CPU and GPU) it takes the same time. (about 200 ms..) It does not seem to be using the GPU... I would like it to use the Intel Gpu to speed things up.. 

In particular the inference time is 7ms, and the parsing time round 200.. for both cases, CPU and GPU

I tried running it like this: 
`python test.py -i test.jpg -m bestf.xml --device GPU`

Running this command: print("Available Devices: ", OpenVinoIE.available_devices)

returns: Available Devices:  ['CPU', 'GPU']

THANK YOU!
